### ISSUE TITLE
Add combineReducers

### ISSUE THREAD

**Issue by stasm** [stasm/innerself#19]
When working on medium-sized apps, it's helpful to enforce the separation of concerns by writing separate reducers for different parts of the state tree. An example of this is routing which needs its own state and logic.

I'd like to add an optional `combineReducers` function to _innerself_. Perhaps in a new module: `innerself/combine`?

@bsouthga created his [own implementation in innerself-hg](https://github.com/bsouthga/innerself-hn/blob/b66b172415ac3f7007194b235a731ad229a2c3c6/src/store/util.ts#L112-L123) and I used [this one](https://github.com/piesku/moment-lost/blob/f0f66bc369a5856d7122da3d81431295d54b55d2/src/store.js#L6-L15) in _A moment lost in time_.

I wonder if we should also add selectors to `connect`. _innerself_ actually [used to have selectors](https://github.com/stasm/innerself/commit/9c5b204bea6199e5dce68a7c2d8289dcc55af17b) when I first wrote it. I then removed them to keep things simple.

Selectors are functions which select a part of the state before it's passed on to the connected component. Something like:

```javascript
const ConnectedFoo = connect(state => state.foo)(FooComponent);
```

In Redux selectors are super-useful: they help Redux decide if the connected component needs to update. In _innerself_ everything re-renders at all times, so the main benefit would be decoupling the shape of the state tree from the API of individual components.

Even though this looks nice and makes the code self-documenting, I expect the most common scenario to not need selectors. We'd then force the `connect(state => state)(…)` boilerplate onto everyone or have a special-case for the undefined selector: `connect()(…)`. This in turn would be different from how Redux treats the undefined selector: it ignores the store completely. Perhaps that's OK. If you want to ignore state, you simply don't `connect` your component to it in _innerself_.

For now in my code I went for the following pattern whenever I needed selectors:

```js
const ConnectedFoo = connect((state, ...args) => FooComponent(state.foo, ...args));
```

This pattern could be easily abstracted to a helper function, for instance:

```js
function select(selector, component) {
    return (state, ...args) => component(selector(state), ...args);
}
```

And then:

```js
const ConnectedFoo = connect(select(state => state.foo, FooComponent))                                                                                                                                                                                      
```

This helper could also live in the optional module alongside `combineReducers`. I'm not particularly fond of this API, though.

**Comment by stasm** [comment]
cc @bsouthga, @flynnham, @bcruddy, @wrick17

**Comment by stasm** [comment]
Another idea: `connect` could take a selector as an optional second argument:

```js
const ConnectedFoo = connect(FooComponent, state => state.foo);
```

**Comment by f1yn** [comment]
While I can't say I'm experienced with Redux (yet), I can say that having the selector as the second argument would be beneficial, since it could denote a simple `state => state` selector as the default (as you already pointed out).

@stasm, to be be clear, the selectors would allow specific renders to happen only when the correct state changes, correct? If so then yeah, I am on-board with this sort of change.

**Comment by wrick17** [comment]
I don't really selectors are going to make much of a difference performance wise. The whole block is re-rendered anyways. However it helps in code sanity, so I think it should be fine as an optional parameter.

Also I was actually thinking about combineReducers this weekend. it would be a great (and necessary) addition. Of course as an additional module.

**Comment by stasm** [comment]
> @stasm, to be be clear, the selectors would allow specific renders to happen only when the correct state changes, correct? If so then yeah, I am on-board with this sort of change.

@wrick17 is right — this won't affect performance. Innerself just re-renders the entire `attach`-ed component every time its output changes.

And I also agree with the point about code sanity. That's what I meant by "self-documenting". Selectors help express in code which part of the state tree the component needs without hard-coding the shape of the tree into the component itself.

**Comment by wrick17** [comment]
@stasm Makes sense. Putting it as an optional parameter sounds good to me. People can choose to write less code or could go for more readable code for bigger projects.

**Comment by bsouthga** [comment]
Adding a `combineReducers` function would be nice, definitely something that will usually be useful.

As for selectors, one option might be to adopt/suggest a `props` pattern similar to react for component arguments, this might make composing components easier (in terms of dealing with an object vs a variable number of arguments), as well as allow for selecting multiple properties like `mapStateToProps`:

```javascript
import html from 'innerself';
import { connect } from '../store';

const ConnectedFoo = connect(
  state => ({ bar: state.innermost.prop.bar }),
  ({ bar }) => html`
    <div>
        check out my bar: ${bar}
    </div>
  `
);
```

(the above would be achievable with your selector proposal in general, just thought it could be a nice pattern to suggest, build around)

**Comment by bsouthga** [comment]
I also like the idea of mirroring `react-redux`'s HOC pattern for `connect` by returning a function which takes a component.

```javascript
const ConnectedFoo = connect(state => ({ bar: state.inner.bar }))(Foo)
```

**Comment by stasm** [comment]
Yeah, I like that too. Redux's `connect` is a configurable decorator which is a really nice design pattern.

Originally, I didn't want to enforce any pattern wrt. `props` vs. arguments passed to components because I didn't want to make the design more complex without a good reason. That reason might have just revealed itself :)

**Comment by stasm** [comment]
What I like about the variable args patterns is that you can then do things like `${tasks.map(ActiveTask)}`.

**Comment by bcruddy** [comment]
Optional parameters or returning a function that accepts a component both seem fine to me with a slight preference towards variable args.

My main concern around all of this is the potential for `innerself` to get away from what it really excels at - being ultra lightweight with minimal setup to put together proofs-of-concept or small personal projects. With its limitations around render speed (e.g. always re-rendering everything) it seems like pushing developers to break apart state in reducers encourages building large, complex applications that will suffer massively in rendering performance. The rendering issue has been solved - if you need high performance rendering react, angular, and vue are much better choices than innerself.

All of that being said - with a router on the horizon, multiple reducers do seem to be the logical next step and the `const ConnectedFoo = connect(state => ({ bar: state.inner.bar }))(Foo)` pattern is already well established in the community.

**Comment by stasm** [comment]
> The rendering issue has been solved - if you need high performance rendering react, angular, and vue are much better choices than innerself.

I completely agree. And I share your concerns—I wouldn't want innerself to get too advanced precisely because I believe that its simplicity communicates its purpose.

@bsouthga has a good point about how this discussion also touches the good practice around the API of components. If the reason to introduce selectors is to improve code sanity and separate the operation of _connecting_ a component from _defining_ it, we shouldn't taint the signature of components with the connected state. The current solution of passing the state as the first argument doesn't scale well: it tight-couples the component's definition with the fact that it is supposed to be connected.

React/Redux have a good solution for this, as @bsouthga already pointed out: the pattern of passing a `props` object as the first argument. I've come to think that _innerself_ should encourage this pattern too. Dumb components don't have to follow it, but the ones which might become connected components should. To ensure components can be re-used, I think it makes sense to establish a good practice of using `props`.

I have a WIP branch locally which I'll push later tonight for your review.

**Comment by stasm** [comment]
Alright, I submitted #22 and would love to hear your thoughts about it!
